### PR TITLE
Function symbol Ruby Wrappers

### PR DESCRIPTION
**Pull request by rajithv**
Wrappers for the function symbol class

### CODE REVIEW

File: ext/symengine/ruby_function.c [hunk 1]
```diff
@@ -37,3 +39,39 @@ IMPLEMENT_ONE_ARG_FUNC(gamma);
 
 #undef IMPLEMENT_ONE_ARG_FUNC
 
+VALUE cfunction_functionsymbol_init(VALUE self, VALUE args)
+{
+    int argc = NUM2INT(rb_funcall(args, rb_intern("length"), 0, NULL));
+    printf("argc : %d\n", argc);
```
Comment (abinashmeher999) [inline ext/symengine/ruby_function.c]: The function shouldn't print anything.

File: ext/symengine/ruby_function.c [hunk 2]
```diff
@@ -37,3 +39,39 @@ IMPLEMENT_ONE_ARG_FUNC(gamma);
...
+    VALUE first = rb_ary_shift(args);
+    if( TYPE(first) != T_STRING ){
+        rb_raise(rb_eTypeError, "String expected as first argument");
```
Comment (abinashmeher999) [inline ext/symengine/ruby_function.c]: Also raise error, if no arguments are passed.

File: ext/symengine/ruby_function.c [hunk 3]
```diff
@@ -37,3 +39,39 @@ IMPLEMENT_ONE_ARG_FUNC(gamma);
...
+    }
+    char *name = StringValueCStr( first );
+    char *c;
+    
+    CVecBasic *cargs = vecbasic_new();
+
+    //basic_struct *cargs[argc];
+
+    basic x;
+    basic y;
```
Comment (abinashmeher999) [inline ext/symengine/ruby_function.c]: This is not needed.

File: ext/symengine/ruby_function.c [hunk 4]
```diff
@@ -37,3 +39,39 @@ IMPLEMENT_ONE_ARG_FUNC(gamma);
...
+    basic_new_stack(x);
+    int i;
+    for(i = 0; i < argc-1; i++){
+        printf("%d : ", i);
```
Comment (abinashmeher999) [inline ext/symengine/ruby_function.c]: Again this should be removed.

File: ext/symengine/ruby_function.c [hunk 5]
```diff
@@ -37,3 +39,39 @@ IMPLEMENT_ONE_ARG_FUNC(gamma);
...
 
+VALUE cfunction_functionsymbol_init(VALUE self, VALUE args)
+{
+    int argc = NUM2INT(rb_funcall(args, rb_intern("length"), 0, NULL));
+    printf("argc : %d\n", argc);
+    VALUE first = rb_ary_shift(args);
+    if( TYPE(first) != T_STRING ){
+        rb_raise(rb_eTypeError, "String expected as first argument");
+    }
+    char *name = StringValueCStr( first );
```
Comment (abinashmeher999) [inline ext/symengine/ruby_function.c]: The spaces around `first` should be removed.

File: ext/symengine/ruby_function.c [hunk 6]
```diff
@@ -37,3 +39,39 @@ IMPLEMENT_ONE_ARG_FUNC(gamma);
 
 #undef IMPLEMENT_ONE_ARG_FUNC
 
+VALUE cfunction_functionsymbol_init(VALUE self, VALUE args)
+{
+    int argc = NUM2INT(rb_funcall(args, rb_intern("length"), 0, NULL));
```
Comment (abinashmeher999) [inline ext/symengine/ruby_function.c]: Or you can use [`rb_ary_length`](http://rxr.whitequark.org/mri/source/array.c#1766).

File: ext/symengine/ruby_function.c [hunk 7]
```diff
@@ -37,3 +39,39 @@ IMPLEMENT_ONE_ARG_FUNC(gamma);
...
+    VALUE first = rb_ary_shift(args);
+    if( TYPE(first) != T_STRING ){
+        rb_raise(rb_eTypeError, "String expected as first argument");
+    }
+    char *name = StringValueCStr( first );
+    char *c;
+    
+    CVecBasic *cargs = vecbasic_new();
+
+    //basic_struct *cargs[argc];
```
Comment (abinashmeher999) [inline ext/symengine/ruby_function.c]: This line should go too.

File: ext/symengine/ruby_function.c [hunk 8]
```diff
@@ -37,3 +39,39 @@ IMPLEMENT_ONE_ARG_FUNC(gamma);
...
+VALUE cfunction_functionsymbol_init(VALUE self, VALUE args)
+{
+    int argc = NUM2INT(rb_funcall(args, rb_intern("length"), 0, NULL));
+    printf("argc : %d\n", argc);
+    VALUE first = rb_ary_shift(args);
+    if( TYPE(first) != T_STRING ){
+        rb_raise(rb_eTypeError, "String expected as first argument");
+    }
+    char *name = StringValueCStr( first );
+    char *c;
```
Comment (abinashmeher999) [inline ext/symengine/ruby_function.c]: This is also not used anywhere I feel.

Comment (rajithv) [inline ext/symengine/ruby_function.c]: @abinashmeher999 all those were used for debugging. I didn't notice that I've added them to the pushed commit. I'll remove them.

File: symengine_version.txt [hunk 9]
```diff
@@ -1,2 +1,3 @@
-b91c64a3b82f15133f4a7df64749523bb0bed185
+1d80d81096b66c08ad3fe4e43ee586cd91b5ccf8
+
```
Comment (isuruf) [inline symengine_version.txt]: No need of these extra lines

File: ext/symengine/ruby_function.c [hunk 10]
```diff
@@ -37,3 +39,36 @@ IMPLEMENT_ONE_ARG_FUNC(gamma);
 
 #undef IMPLEMENT_ONE_ARG_FUNC
 
+VALUE cfunction_functionsymbol_init(VALUE self, VALUE args)
+{
+    int argc = RARRAY_LEN(args);
+    if(argc == 0){
```
Comment (isuruf) [inline ext/symengine/ruby_function.c]: `if (argc == 0) {`

File: ext/symengine/ruby_function.c [hunk 11]
```diff
@@ -37,3 +39,36 @@ IMPLEMENT_ONE_ARG_FUNC(gamma);
...
+        rb_raise(rb_eTypeError, "Arguments Expected");
+    }
+
+    VALUE first = rb_ary_shift(args);
+    if( TYPE(first) != T_STRING ){
```
Comment (isuruf) [inline ext/symengine/ruby_function.c]: `if (TYPE(first) != T_STRING) {`

File: ext/symengine/ruby_function.c [hunk 12]
```diff
@@ -37,3 +39,36 @@ IMPLEMENT_ONE_ARG_FUNC(gamma);
...
+        rb_raise(rb_eTypeError, "String expected as first argument");
+    }
+    char *name = StringValueCStr(first);
+    
+    CVecBasic *cargs = vecbasic_new();
+
+    basic x;
+    basic_new_stack(x);
+    int i;
+    for(i = 0; i < argc-1; i++){
```
Comment (isuruf) [inline ext/symengine/ruby_function.c]: Space after `for` and before `{`.

**Comment by isuruf** [comment]
Python wrappers have a nice syntax for creating FunctionSymbols.

`func = Function('f')`
Above would create an UndefFunction object which is a Python class, and 
`func(x, 2*y)` would create a FunctionSymbol `f(x, 2*y)`.

What do you think?

**Comment by rajithv** [comment]
@isuruf What are the advantages of doing it in two steps?

What about, haveing `SymEngine::Function('f', x, 2*y)` as an alias for `SymEngine::FunctionSymbol.new(...)` ?

File: spec/function_symbol_spec.rb [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+  end
+  
+  context '#diff' do
+    let(:fun) { (SymEngine::FunctionSymbol.new('f', 2*x, y, SymEngine::sin(z))) }
+    context 'by variable' do
+        subject { fun.diff(x)/2 }
+        it { is_expected.to be_a SymEngine::Subs }
+      end
+
+      context 'by constant' do
```
Comment (isuruf) [inline spec/function_symbol_spec.rb]: Remove this. Differentiation should throw an error if the input is not a Symbol.

File: spec/function_symbol_spec.rb [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+      it { is_expected.to be_a SymEngine::FunctionSymbol }
+    end
+  end
+  
+  context '#diff' do
+    let(:fun) { (SymEngine::FunctionSymbol.new('f', 2*x, y, SymEngine::sin(z))) }
+    context 'by variable' do
+        subject { fun.diff(x)/2 }
+        it { is_expected.to be_a SymEngine::Subs }
+      end
```
Comment (isuruf) [inline spec/function_symbol_spec.rb]: Above three lines have to be shifted to the left by 2 spaces

**Comment by isuruf** [comment]
Advantage is for easy typing when there are multiple function symbols made and also for readability

```
f = Function('f')
g = f(x, y) - 2 * f(y - x, x)
```
instead of

```
g = FunctionSymbol.new("f", x, y) - 2 * FunctionSymbol.new("f", y - x, x)
```

File: lib/symengine.rb [hunk 15]
```diff
@@ -29,6 +28,10 @@ def symbols ary_or_string, *params
       end
     end
   end
+  
+  def SymEngine.Function(n)
```
Comment (isuruf) [inline lib/symengine.rb]: You can move this inside `class << self` block and then you can remove `SymEngine.`

**Pull request merged**